pkg/minikube/service/service_test.go

```diff
@@ -493,4 +538,341 @@ func TestGetServiceURLsForService(t *testing.T) {
+		secret := "mock-secret"
+		K8s = &MockClientGetter{
+			servicesMap:  serviceNamespaces,
+			endpointsMap: endpointNamespaces,
+			secretsMap:   secretsNamespaces,
+		}
+		defer revertK8sClient(K8s)
+		if err := DeleteSecret(ns, secret); err == nil {
+			t.Fatalf("DeleteSecret expected to fail for namespace: %v which doesn't exist", ns)
+		}
+	})
+}
+
+func TestCreateSecret(t *testing.T) {
+	t.Run("failed Get Client", func(t *testing.T) {
+		ns := "foo"
+		K8s = &MockClientGetter{
+			servicesMap:  serviceNamespaces,
+			endpointsMap: endpointNamespaces,
+			secretsMap:   secretsNamespaces,
+		}
+		getCoreClientFail = true
+		defer revertK8sClient(K8s)
+		if err := CreateSecret(ns, "foo", map[string]string{"ns": "secret"}, map[string]string{"ns": "baz"}); err == nil {
+			t.Fatal("CreateSecret expected to fail for failed GetCoreClient")
+		}
+	})
+	t.Run("bad namespace", func(t *testing.T) {
+		ns := "no-such-namespace"
+		K8s = &MockClientGetter{
+			servicesMap:  serviceNamespaces,
+			endpointsMap: endpointNamespaces,
+			secretsMap:   secretsNamespaces,
+		}
+		defer revertK8sClient(K8s)
+		if err := CreateSecret(ns, "foo", map[string]string{"ns": "secret"}, map[string]string{"ns": "baz"}); err == nil {
+			t.Fatalf("CreateSecret expected to fail for namespace: %v", ns)
+		}
+	})
+	t.Run("ok", func(t *testing.T) {
+		ns := "foo"
+		K8s = &MockClientGetter{
+			servicesMap:  serviceNamespaces,
+			endpointsMap: endpointNamespaces,
+			secretsMap:   secretsNamespaces,
+		}
+		defer revertK8sClient(K8s)
+		if err := CreateSecret(ns, "foo", map[string]string{"ns": "secret"}, map[string]string{"ns": "baz"}); err != nil {
+			t.Fatalf("CreateSecret not expected to fail but got: %v", err)
+		}
+	})
+}
+
+func TestWaitAndMaybeOpenService(t *testing.T) {
+	defaultAPI := &tests.MockAPI{
+		FakeStore: tests.FakeStore{
+			Hosts: map[string]*host.Host{
+				config.GetMachineName(): {
+					Name:   config.GetMachineName(),
+					Driver: &tests.MockDriver{},
+				},
+			},
+		},
+	}
+	defaultTemplate := template.Must(template.New("svc-template").Parse("http://{{.IP}}:{{.Port}}"))
+
+	var tests = []struct {
+		description string
+		api         libmachine.API
+		namespace   string
+		service     string
+		expected    []string
+		err         bool
+	}{
+		{
+			description: "no host",
+			api: &tests.MockAPI{
+				FakeStore: tests.FakeStore{
+					Hosts: make(map[string]*host.Host),
+				},
+			},
+			err: true,
+		},
+		{
+			description: "correctly return serviceURLs",
+			namespace:   "default",
+			service:     "mock-dashboard",
+			api:         defaultAPI,
+			expected:    []string{"http://127.0.0.1:1111", "http://127.0.0.1:2222"},
+		},
+		{
+			description: "correctly return empty serviceURLs",
+			namespace:   "default",
+			service:     "mock-dashboard-no-ports",
+			api:         defaultAPI,
+			expected:    []string{},
+			err:         true,
+		},
+	}
+	defer revertK8sClient(K8s)
+	for _, mode := range []bool{true, false} {
```

Comment:
This nested `for` test is really hard to read. Can you break it into an array of structures, and then do only one `for` to test it?

Reply:
fixed. Should be better now.